Raft test topology

alecco opened the pull request:
Give cluster control to pytests. While there add missing stop gracefully and add server to ScyllaCluster.

Clusters can be marked dirty but they are not recycled yet. This will be done in a later series.

Commits are

- test.py: ScyllaServer stop gracefully
- test.py: ScyllaCluster add_server helper
- test.py: harness for cluster management from test

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/895/

kostja commented:
Overall looks very powerful. It's amazing how little code it is given into account it replaces dtest, ccm, nodetool and jmx combined. Way to go!

kostja commented on test.py:
```diff
@@ -702,6 +703,7 @@ def print_summary(self):
     async def run(self, options):
         async with self.suite.clusters.instance() as cluster:
             self.args.insert(0, "--host={}".format(cluster[0].host))
+            self.args.append(f"--cluster_sock={cluster.sock_path}")
```
Let's see if we could get rid of _

kostja commented on test/pylib/cluster_cli.py:
```diff
@@ -0,0 +1,68 @@
+
+
+class ClusterCli():
+    def __init__(self, sock_path: str):
```
Please provide comments and copyright

kostja commented on test/pylib/schema_helper.py:
```diff
@@ -0,0 +1,299 @@
+    from cassandra.cluster import Session as CassandraSession            # type: ignore
+
+
+new_keyspace_id = itertools.count(start=1).__next__
```
Let's use unique_name()?

alecco replied:
I don't like `unique_name()` as it's ugly, hard to repro (takes from timestamp), and hacky.

An itertools generator is easier to read, easier to repro, and even thread safe.

kostja replied:
Fine, let's put your logic into unique_name() then. My only point is that we don't have two ways of making names. We could make the names randomness depends on a single rnd seed, too.

alecco replied:
Removed this to use the keyspace fixture

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/996/

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -337,6 +337,27 @@ async def stop(self) -> None:
+            if self.cmd:
+                logging.info("stopped server at host %s", hostname)
+            self.cmd = None
+            self.control_connection = None
```
This needs to be updated to shutdown the control cluster (take a look at stop() in the current master.

alecco replied:
rebased

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -337,6 +337,27 @@ async def stop(self) -> None:
             self.control_connection = None
             self.control_cluster = None
 
+    async def stop_gracefully(self) -> None:
```
I wonder if we need a timeout param here.

alecco replied:
Added to TODO list

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -387,20 +387,24 @@ def __init__(self, replicas: int,
+    async def add_server(self):
+        server = self.create_server(self.name, self.last_seed)
+        await server.install_and_start()
+        self.cluster[server.host] = server
```
self.cluster has an int key, did I miss the change to str keys?

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -379,39 +382,68 @@ def write_log_marker(self, msg) -> None:
 
     async def install_and_start(self) -> None:
         try:
+            seed = None
```
Unused?

alecco replied:
Resolved in latest

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1077/

alecco commented on test/pylib/scylla_server.py:
```diff
@@ -410,5 +465,124 @@ def after_test(self, name) -> None:
+        return aiohttp.web.Response(text=f"{','.join(sorted(self.started.keys()))}")
+
+    async def cluster_stop(self, request):
+        await asyncio.gather(*(server.stop() for server in self.started.values()))
```
addressed in v6

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1115/

alecco commented on test/pylib/scylla_server.py:
```diff
@@ -410,5 +465,124 @@ def after_test(self, name) -> None:
+        return aiohttp.web.Response(text="OK")
+
+    async def cluster_node_stop(self, request):
+        node_id = request.match_info['id']
```
Return HTTP 500

alecco commented:
Please ignore the last couple of force pushes. I'm still working on the review and will send another version.

alecco commented on test/pylib/scylla_server.py:
```diff
@@ -387,20 +387,24 @@ def __init__(self, replicas: int,
 
+    async def add_server(self):
+        server = self.create_server(self.name, self.last_seed)
+        await server.install_and_start()
```
Harness needs lookup by host id. We need to get it before adding to new  `cluster.started[host_id] = server`

alecco replied:
But if you can figure out a better way, I'm all ears.

alecco commented on test/pylib/cluster_cli.py:
```diff
@@ -0,0 +1,72 @@
+        self.conn = aiohttp.UnixConnector(path=self.sock_path)
+        self.session = aiohttp.ClientSession(connector=self.conn)
+
+    async def stop(self):
```
`harness_cli.stop()`  was removed as `Harness` instance lives for the whole pytest so test cases shouldn't stop it.

alecco commented on test/pylib/scylla_server.py:
```diff
@@ -358,35 +382,66 @@ def write_log_marker(self, msg) -> None:
+        self.app = aiohttp.web.Application()
+        self.setup_routes()
+        self.runner = aiohttp.web.AppRunner(self.app)
+        self.sock_path = f"{tmpdir}/harness_sock_{randint(1000000,9999999)}"
```
Now it's using a safe temporary directory and the sock inside of it. It's the pattern in Python as there is no possible way to create safe unique sock filenames.

alecco commented:
CQL discovery patch has its own pull  https://github.com/scylladb/scylla/pull/11124

alecco commented:
CQL log handling has its own pull https://github.com/scylladb/scylla/pull/11125

kostja commented:
@nyh @kbr- there's been a lot of changes in this pull request, could you please review?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1499/

nyh commented on test/conftest.py:
```diff
@@ -99,21 +99,30 @@ def cluster_con(hosts: List[str], port: int):
 # "cql" fixture: set up client object for communicating with the CQL API.
 # We use scope="session" so that all tests will reuse the same client object.
-@pytest.fixture(scope="session")
+@pytest.fixture(scope="function")
```
This (making a fixture function instead of session) is a big change, which makes some sense for other reasons (see https://github.com/scylladb/scylla/issues/9599) but I don't understand the reason here and if we don't do too much work on each and every test.
But as long as this doesn't touch cql-pytest and slow down all its quick tests, I won't fight it.

alecco replied:
For topology tests, clusters are used per test _case_. So cql connection needs to be updated. I thought it's overkill to add logic to keep the driver connected when almost all cases will use a fresh cluster.

BTW in the refactor I forgot to update the comment to reflect this.

alecco replied:
Now this lives in `test/topology/conftest.py` and nothing else is affected. All topology tests need the driver connection to be managed per test case.

nyh commented on test/conftest.py:
```diff
@@ -139,7 +148,7 @@ def cql_test_connection(cql, request):
 # a "this_dc" fixture to figure out the name of the current DC, so it can be
 # used in NetworkTopologyStrategy.
-@pytest.fixture(scope="session")
+@pytest.fixture(scope="function")
```
Again, you're giving up on a performance advantage here (don't need to figure out the name of the DC in each separate test), and I didn't understand why.

alecco replied:
I'll try to cache this.

alecco replied:
Obsolete in latest. Topology doesn't need `this_dc` hack as per other comment.

nyh commented on test/conftest.py:
```diff
@@ -167,9 +177,9 @@ async def fails_without_raft(request, check_pre_raft):
-@pytest.fixture(scope="session")
+# and automatically deleted at the end.
+@pytest.mark.asyncio
+@pytest.fixture(scope="function")
```
This change is even bigger. It means a new keyspace is created for every test.
Again, I'm fine with this is that's what you want, but as long as it doesn't read cql-pytest.

nyh replied:
Please note that because each keyspace will have a unique name, this will suffer from the still-existing bug that Scylla has when you create and delete a lot of keyspaces or tables with different names.

alecco replied:
A new cluster is created for almost every topology test case. Same reasoning as driver connection.

nyh commented on test/pylib/harness_cli.py:
```diff
@@ -125,3 +125,10 @@ async def nodes(self) -> List[str]:
+        """Manually mark current cluster dirty.
+           To be used when a server was modified outside of this API."""
+        self.dirty = True
+        await self._request_and_check("http://localhost/cluster/mark-dirty",
```
Again, why "localhost"? Shouldn't this be the Unix-domain socket filename?

alecco replied:
As before, `aiohttp`.

nyh reviewed (changes requested):
I have a lot of small comments and questions, nothing major, but marking "request changes" because of specific places that I don't want you to break  test/cql-pytest.

You can run "test/cql-pytest/run |& less" to verify that you didn't break the other ways of running it (test/cql-pytest/run, test/cql-pytest/run-cassandra, and just "pytest" should all work and we don't want to break them).

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -364,6 +364,27 @@ async def stop(self) -> None:
+        except ProcessLookupError:
+            pass
+        else:
+            await self.cmd.wait()
```
@nyh in general I agree, especially since this has now an exposure in the test itself. there is an overall timeout on running tests in jenkins, so I thought it's not urgent and not requested in my review.

alecco replied:
This is pre-existing. I'll address later.

kostja commented on test.py:
```diff
@@ -34,7 +34,7 @@
 from test.pylib.host_registry import HostRegistry
 from test.pylib.pool import Pool
-from test.pylib.scylla_server import ScyllaServer, ScyllaCluster
+from test.pylib.scylla_cluster import ScyllaCluster, Harness
```
Would sidecar be better? The idea is that it's a separate entity to manage the cluster, not the cluster itself.

alecco replied:
ClusterManager ClusterAdministrator ClusterController ClusterDirector ClusterRunner ClusterConductor ClusterDriver
?

alecco replied:
Lacking answers, `ScyllaClusterManager` it is.

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -531,8 +550,87 @@ def after_test(self, name) -> None:
+        if server_id in self.stopped or server_id in self.removed:
+            return True
+        if server_id not in self.started:
+            return False
```
no, why, the test can do random things, including this one. we can inform the test that it's trying to stop a stopped instance over rest and it will decide what to do with it.

alecco replied:
Improved the logic and the result is now [success:bool, msg:str]

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -531,8 +550,87 @@ def after_test(self, name) -> None:
+            return False
+        await server.stop_gracefully()
+        await server.start()
+        return True
```
We intend to test exactly this eventually. It is expected to work (or will be expected to work when we test it - I can't guarantee the test will pass today of course).

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -531,8 +550,87 @@ def after_test(self, name) -> None:
+        return True
+
+    async def node_remove(self, server_id: str) -> bool:
+        """Remove a specified server"""
```
This part is incomplete. I agree it should call removenode rest api, not stop/uninstall the node.

kostja replied:
It will be addressed in a follow up.

alecco replied:
I'll remove this for now

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -531,8 +550,87 @@ def after_test(self, name) -> None:
+        return True
+
+    async def node_replace(self, old_node_id) -> Optional[str]:
+        """Replace a specified server with a new one"""
```
Same here.

alecco replied:
removed

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -493,7 +493,10 @@ async def stop_gracefully(self) -> None:
+    async def add_server(self, mark_dirty:bool=True) -> str:
+        """Add a new server to the cluster"""
+        if mark_dirty:
+            self.is_dirty = True
```
I think you're missing the purpose of set_dirty: the idea is that every cluster that was manipulated with via rest api is "dirty", and every cluster that wasn't isn't, so it can be reused for the next test.

alecco replied:
Refactored as Nadav requested.

kostja commented on test/pylib/harness_cli.py:
```diff
@@ -0,0 +1,84 @@
+
+    async def is_harness_up(self) -> bool:
+        """Check if Harness server is up"""
+        ret = await self._request_and_check("http://localhost/up",
```
It's a conventional host name which should be used for asyncio http over unix domain socket.

alecco replied:
`aiohttp` requires a host defined  https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client_reqrep.py#L308

alecco commented on test.py:
```diff
@@ -339,6 +339,9 @@ def topology_for_class(self, class_name: str, cfg: dict) -> Callable[[], Awaitab
         def create_server(cluster_name, seed):
             cmdline_options = self.cfg.get("extra_scylla_cmdline_options", [])
+            config_options = self.cfg.get("extra_scylla_config_options", 
+                                          {"authenticator": "AllowAllAuthenticator",
```
I reconsidered and will leave with the previous default and instead have topology tests set `AllowAll*` in `suite.yaml`. This way I don't modify cql-pytest nor rest-api.

alecco commented on test/pylib/scylla_server.py:
```diff
@@ -28,6 +28,9 @@
 developer_mode: true
 
+# Port for the CQL native transport to listen for clients on
+native_transport_port: {cql_port}
```
As there was a --port option everywhere I wanted to have Harness to match this and provide host/port. But I can just remove this and hardcode it to 9042.

alecco replied:
Back to leaving as it was. But then the `--port` option in pytests is pointless.

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -431,7 +432,11 @@ def __init__(self, scylla_exe: str, replicas: int, test_base_dir: str,
         self.cluster: List[ScyllaServer] = []
+        self.started: Dict[str, ScyllaServer] = {}
+        self.stopped: Dict[str, ScyllaServer] = {}
+        self.removed: Set[str] = set()
```
Stopped is a server that is there and can be restarted. Removed servers are gone. I planned to give a different error to tests if they try to start a removed server (instead of just a generic error for an unrecognized id). But I forgot to include this. I'll add it in the next version.

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -487,27 +503,30 @@ async def add_server(self) -> None:
     def __getitem__(self, i: int) -> ScyllaServer:
-        return self.cluster[i]
+        assert i >= 0, "ScyllaCluster: cluster sub-index must be positive"
+        return next(server for pos, server in enumerate(self.started.values()) if pos == i)
```
It's a Python pattern to avoid creating a full list. I can remove it.

alecco replied:
More precisely  dict.values() is a set-like object. https://stackoverflow.com/a/33674108
I'll make it a list and subscript.

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -431,7 +432,11 @@ def __init__(self, scylla_exe: str, replicas: int, test_base_dir: str,
         self.host_registry = host_registry
         self.artifact_registry = artifact_registry
         self.cluster: List[ScyllaServer] = []
+        self.started: Dict[str, ScyllaServer] = {}
```
Forgot to remove `self.cluster` on refactor.

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -638,17 +639,28 @@ async def node_replace(self, old_node_id) -> Optional[str]:
         self.is_after_test_ok: bool = False
         self.is_running: bool = False
+        # API
+        # NOTE: need to make a safe temp dir as tempfile can't make a safe temp sock name
```
Clusters and their Pool exist outside a test suite. Harness "borrows" clusters and once it's done, if the cluster is not dirty, returns them to the Pool. Clusters have their own directory where their servers put files. Harness socket can't (and shouldn't) be inside any given cluster. And since there's no way to safe socket name in Python, the pattern is to create a safe directory and put the socket in there, instead.

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -490,3 +532,46 @@ def after_test(self, name) -> None:
+
+
+class Harness:
+    """Manages a pool of Scylla clusters for running test cases"""
```
@alecco after the latest refactoring the harness doesn't manage the pool any more, let's update the comment. How about "Manages use of scylla clusters by a specific test. Takes the clusters from a pool shared among all harness instances of this test suite. When a test "breaks" the cluster, e.g. downs a node in it, destroys the cluster and takes another one for the pool for the next test, so that one test has no impact on the other"

alecco replied:
> Again, I think the word "Harness" is too generic. If this is a "pool of Scylla clusters" why not call it ScyllaClusterPool or something like that?

ScyllaClusterManager (Pool is at Suite level now, as Kostja points out)

kostja commented on test.py:
```diff
@@ -626,68 +624,64 @@ def _reset(self) -> None:
+            await self.harness.start()
+            self.args.insert(1, f"--harness-api={self.harness.sock_path}")
+            self.is_executed_ok = await run_test(self, options, env=self.env)
+            await self.harness.stop()
```
stop() is not executed if run_test throws. The cluster itself is protected by an artifact, but it still doesn't look safe, e.g. if we retry a flaky test, we will leak clusters in the process. I think an equivalent change is to replace async with clusters.instance() with async iwth harness.instance(). That would preserve the alignment of the code, simplifying the review. What do you think?

alecco replied:
Can't it just be called in `finally:` ?

alecco replied:
Calling in `finally:`

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -0,0 +1,792 @@
+        logging.info("Uninstalling cluster")
+        await self.stop()
+        await asyncio.gather(*(server.uninstall() for server in self.stopped.values()))
+        # Note: self.cluster_dir is removed by suite when all tests pass
```
Shouldn't you delete the cluster directory now that you have it?

alecco replied:
No more cluster directory

kostja commented on test.py:
```diff
@@ -337,32 +337,28 @@ def __init__(self, path, cfg: dict, options: argparse.Namespace, mode: str) -> N
+                                        config_options,
+                                        self.hosts,
+                                        self.artifacts)
                 await cluster.install_and_start()
```
Previously, cluster artifacts were added before install_and_start(), which can easily fail. Now they are added after. I think we should not leak clusters in any cases. Unless there is a strong reason to do otherwise, I suggest artifacts are added before the resources they protect are allocated.

alecco replied:
Reverted to previous Cluster and Server logic.

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -0,0 +1,792 @@
+        logging.info("Cluster %s restarting server %s", self, server_id)
+        if server_id in self.running:
+            self.is_dirty = True
+            server = self.running.pop(server_id)
```
here and elsewhere in this class it bothers me that we may leak the server if the line below throws (stop_gracefully(), start()). This is a rest api call, and it is legitimate to fail, the failure may be well ignored by the test, or the test may be testing for failure. I think we should not leak servers in any case, from any of this list. Do we really need these lists, maybe the server internal 'state' variable (running, stopped, removed) would be enough?

alecco replied:
API commands need to get a server by id so having a plain list and state inside the Server object is not good enough.

It's true this might leak. I'll fix later.

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -0,0 +1,792 @@
+        self.is_running: bool = False
+        # API
+        # NOTE: need to make a safe temp dir as tempfile can't make a safe temp sock name
+        self.harness_dir: str = tempfile.mkdtemp(prefix="harness-", dir=base_dir)
```
Now that we create an instance of harness for each test, to simplify debugging it would be best in my view to derive the name of the socket from the test name. something like test-name.test-id.sock.
Please note that this file needs to be removed in the end, so please create an artifact for it.

alecco replied:
As topology tests require Scylla and Raft, I removed the `this_dc` hack from topology and use normal replication for keyspaces there. Other test suites are unchanged.

kostja commented on test/pylib/harness_cli.py:
```diff
@@ -0,0 +1,168 @@
+
+    async def before_test(self, test_name: str) -> None:
+        """Before a test starts check if cluster needs cycling and update driver connection"""
+        dirty = await self.is_dirty()
```
I don't quite understand the timing of this. This is a check that runs before test. And it runs an rpc against harness which disposes of a dirty cluster *after* the test. So if the cluster was dirty and was disposed of, this rpc apparently woudl return false, which is wrong. Isn't it safer to use some kind of cluster uuid, and check if it has changed for whatever reason?

kostja replied:
In fact we already have the cluster uuid...

alecco replied:
Cluster is disposed in before test. Disposing in after test doesn't make sense, else if last test makes the cluster dirty, manager will bring up a new cluster for nothing (no next test).

kostja commented on test/pylib/scylla_cluster.py:
```diff
@@ -0,0 +1,792 @@
+        self.app.router.add_get('/cluster/servers', self._cluster_servers)
+        self.app.router.add_get('/cluster/before_test/{test_name}', self._before_test_req)
+        self.app.router.add_get('/cluster/after_test/{test_name}', self._after_test)
+        self.app.router.add_get('/cluster/mark-dirty', self._mark_dirty)
```
I think we agreed to follow _ vs - rules consistently, and settled on underscores. Please rename is-dirty -> is_dirty, mark-dirty -> mark_dirty. Thanks.

alecco replied:
This was a mistake. URL should've been `/before-test`  `/after-test`

kostja commented on test/pylib/harness_cli.py:
```diff
@@ -0,0 +1,168 @@
+    async def mark_dirty(self) -> None:
+        """Manually mark current cluster dirty.
+           To be used when a server was modified outside of this API."""
+        self.dirty = True
```
could you please explain why you need self.dirty on the client? Shouldn't you trust the server?

alecco replied:
I don't, will remove

alecco commented on test/conftest.py:
```diff
@@ -18,7 +18,6 @@
 from cassandra.policies import RoundRobinPolicy                          # type: ignore
 from test.pylib.util import unique_name                                  # type: ignore
 import pytest
-import ssl
```
OK removed in latest

alecco commented on test/pylib/scylla_cluster.py:
```diff
@@ -531,8 +550,87 @@ def after_test(self, name) -> None:
+        """Start a stopped node"""
+        self.is_dirty = True
+        logging.info("Cluster %s starting server", self)
+        server = self.stopped.pop(server_id, None)
```
Fixed in latest

alecco commented on test/cql-pytest/conftest.py:
```diff
@@ -27,6 +27,8 @@ def pytest_addoption(parser):
     parser.addoption('--ssl', action='store_true',
         help='Connect to CQL via an encrypted TLSv1.2 connection')
+    parser.addoption('--harness-api', action='store', required=True,
+                     help='Harness unix socket path')
```
Removed.

But now I'm not happy there's `--host` and `--port` left at `test/conftest.py` for all suites where topology tests don't need nor use them.

alecco commented on test/rest_api/conftest.py:
```diff
@@ -34,6 +34,8 @@ def pytest_addoption(parser):
         help='Connect to CQL via an encrypted TLSv1.2 connection')
     parser.addoption('--api-port', action='store', default='10000',
         help='server REST API port to connect to')
+    parser.addoption('--harness-api', action='store', required=True,
```
Reverted

alecco commented:
Main changes in latest version:

- splits logic in `test/conftest.py` to `test/topology/conftest.py`
- `Harness` renamed `ScyllaClusterManager`
- Created `TopologyTestSuite` and `TopologyTest` to isolate behaviour from other test suites
- Previous 3 make code not need to modify other suites
- addressed all reviews
- except concern for leaks, timeout for stops, locking API async actions (just added a comment for now as Nadav requested)

alecco commented:
Manual CI https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1611/
(didn't start automatically due to repo name change)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1613/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1615/

kbr- commented on test/pylib/scylla_cluster.py:
```diff
@@ -483,3 +484,29 @@ def after_test(self, name) -> None:
+        self.test_name: str = test_name
+        self.clusters: Pool[ScyllaCluster] = clusters
+
+    async def start(self) -> None:
```
>  Still, borrowing/getting a server could have a better API

Or maybe just starting it:
```
with self.manager.start():
    ... use self.manager.cluster ...
```
But it seems that combining start + borrow cluster makes sense.

alecco commented on test/conftest.py:
```diff
@@ -137,45 +93,14 @@ def cql_test_connection(cql, request):
 # tests will reuse the same keyspace.
 @pytest.fixture(scope="session")
-async def keyspace(cql, this_dc):
+def keyspace(cql, this_dc):
```
Nadav wasn't happy with us adding topology-only things to `test/conftest.py` and since nobody besides topology uses async, we can just keep it within `test/topology/conftest.py`. Also, Python driver doesn't have async support for connection. And the `this_dc` workaround is for running tests with Cassandra 3 (while topology tests require Scylla 5+).

I'll update the commit comment.

alecco commented on test/topology/conftest.py:
```diff
@@ -88,3 +89,15 @@ async def random_tables(request, cql, keyspace) -> AsyncGenerator:
+# and automatically deleted at the end. We use scope="session" so that all
+# tests will reuse the same keyspace.
+@pytest.fixture(scope="session")
+def keyspace(cql):
```
Should've read `this_dc` workaround. I'll fix.

alecco commented on test/pylib/random_tables.py:
```diff
@@ -240,6 +240,8 @@ def __init__(self, test_name: str, cql: CassandraSession, keyspace: str):
         self.keyspace = keyspace
         self.tables: List[RandomTable] = []
         self.removed_tables: List[RandomTable] = []
+        cql.execute(f"CREATE KEYSPACE {keyspace} WITH REPLICATION = "
```
It's only used by topology tests for now. If other tests want to use it they can just remove the keyspace fixture dep. It's not auto.

alecco commented:
> Are you planning to address this in this PR or in a follow-up?
> (it's fine for me if it's done in a follow-up)

It's pre-existing. We can address it later. @kbr-

alecco commented on test/cql/suite.yaml:
```diff
@@ -1,3 +1,6 @@
 type: Approval
+extra_scylla_config_options:
+    authenticator: AllowAllAuthenticator
```
As per commit message, the `auth` keyspace starts with `RF=1` and if we are stopping nodes this can cause failures. For now topology tests don't need `auth` so I just disabled it. If in the future topology tests need `auth`, we can switch back to normal auth and we will have to do `ALTER KEYSPACE ... RF = 3` and run repair and wait for it to finish. But this will bring slowdown so we might want to split it.

> but here you also change it for cql suite.

Ouch, I could swear I removed it. Will do next.

alecco commented on test/topology/conftest.py:
```diff
@@ -58,6 +61,34 @@ def run_async(self, *args, **kwargs) -> asyncio.Future:
+# --port options, and defaults to localhost and 9042, respectively.
+# We use scope="session" so that all tests will reuse the same client object.
+@pytest.fixture(scope="session")
+def cql(request):
```
it's not, will fix

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1664/

alecco commented:
In latest push I added `pytest_asyncio` mode auto. Removed leftover `pytest.mark.asyncio`.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1673/

alecco commented:
Changes in latest push:
- Rebase on latest https://github.com/scylladb/scylladb/pull/11193 file and split part 1 pull request https://github.com/scylladb/scylladb/pull/11210
- Use an asynchronous context manager for ClusterManager  (Kamil)
- Untangle wrongly squashed commits in previous push
- Minor fixes

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1685/

nyh commented on test/pylib/manager_client.py:
```diff
@@ -58,6 +58,21 @@ def _driver_update(self) -> None:
+        dirty = await self.is_dirty()
+        if dirty:
+            self.driver_close()  # Close driver connection to old cluster
+        await self._request(f"http://localhost/cluster/before-test/{test_name}")
```
I think I asked this before but can't find my question (or answer) - what is this "localhost" about? Why is this localhost and not a very specific IP address, which we need to allow multiple instances of this test to run in parallel?

nyh replied:
> Why do you use "localhost" when communicating with "the harness"?
> 
> I think we should always think about the possibility that the same test might be run in parallel, and allow everything in it, including the harness, to use a unique IP address or a unique port. 

On second thought, I just remembered that you claimed that you were using a UNIX-domain socket - i.e., a unique temporary file - and not TCP. So what is that "localhost" URLs I saw all over the code? How do you get those "localhost" URLs not to communicate with 127.0.0.1 and instead communicate with a UNIX-domain socket?

kostja replied:
@nyh I haven't tested, but I guess pretty much any domain name would work. What would you like it to be?

kostja replied:
https://github.com/scylladb/scylladb/pull/10815/files/1bfb80b8bb534238f7f40399c4657b1a60a4663f..8a0345250f901a8830a4a400b73272d34a5afdcb#diff-ac4b42f4b43b922bee79039882129028d018615824be306f1845bfdb86b350d2R40

kostja replied:
self.conn = aiohttp.UnixConnector(path=self.sock_path)

nyh replied:
Oh, I understand now. So the hostname is just ignored (and probably gets sent as a "Host:" header if it uses the real HTTP protocol). Maybe it would have been clearer if we had a silly-looking string like "unixsocket" instead of something which looks like it might be real like "localhost". But you're right, I'm happy that localhost is **not** used - what the request looks like is of secondary importance.

nyh reviewed (changes requested):
Why do you use "localhost" when communicating with "the harness"?

I think we should always think about the possibility that the same test might be run in parallel, and allow everything in it, including the harness, to use a unique IP address or a unique port. Even if you claim there there is just one harness - that **it** will in turn run multiple sub-tests in parallel - I still claim that it is possible that on the same test machine, someone will want to run multiple harnesses in parallel. E.g., maybe somebody will run tests for two different versions of Scylla in parallel? Maybe two different users on the same machine (remember that old concept? :-)) might want to run tests in parallel?

I'm not enthusiastic about the whole idea of a "harness" that communicates via HTTP. I don't think it is actually necessary and would have done this without a server. However, if that's the design we chose, I will not oppose it. I just don't want it to introduce new limitations - such as the inability to run two independent test runs in parallel.

nyh commented:
Also I got completely confused what's the difference between this PR, and https://github.com/scylladb/scylladb/pull/11266. Both are open and seem to be actively worked on, and they both seem to do similar things? I'm reviewing both, but I don't understand why.

alecco commented:
> Also I got completely confused what's the difference between this PR, and #11266. Both are open and seem to be actively worked on, and they both seem to do similar things? I'm reviewing both, but I don't understand why.

I was asked to split this pull request in smaller parts.

> This is the third part out of https://github.com/scylladb/scylladb/pull/10815

 I kept this pull to have the older reviews available for reference.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1816/
